### PR TITLE
BUG: Surviving Data in viewer must be visible

### LINKED ISSUE #1107: blinking then removing top layer results in blank viewer
**Reporter: Kyle Conroy**

steps to reproduce:
 * load imviz example
 * blink once
 * uncheck first entry in data menu (viewer becomes blank since all remaining layers were set to visible=False by the blinking)
 * blink again (still blank, because now it only sees on layer, even though its a non-visible layer)
 * re-check data entry (image comes back, can blink again).

expected behavior:
 * unclear. 
 * We could listen to RemoveDataMessage and if it was the active layer, reset the blinking?). 
 * For the case where there is only one layer, blinking should at least ensure that is visible.

related to all of this: blinking edits the `visible` property on the layers, which will then override any manual changes to `visible` by the user.  Is this expected?  Or should we somehow remember the user-set values for `visible` and not include those that are disabled in the blinking cycle?

https://user-images.githubusercontent.com/877591/155400169-bb30cc55-d5b9-4caf-b826-9e721ca1d4ba.mov

[🐱](https://jira.stsci.edu/browse/JDAT-2168)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to fix a bug where data set to not visible after blinking remains invisible even if it is the only Data left in the viewer.

Fixes #1107 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/app.py
Comment (kecnry) [inline jdaviz/app.py:1171]: This does fix the case where all but one entry are unchecked in the data menu, but I can still reproduce the bug if there were three original data layers and then removing whichever is the currently "visible through blinking" (leaving two checked, but hidden by blinking).

Comment (pllim) [reply jdaviz/app.py:1171]: Ah, in that case, this is not the right fix. I need to ponder more. Thanks for pointing that out!

**Comment by pllim** [comment]
Re: Compass -- What it says is can be wrong due to #1126 . I am not sure how to fix that yet. I'll ponder more.

**Comment by pllim** [comment]
@kecnry , I changed the fix but it is still at the same place in `app.py` (plus some extra stuff). Does this work for you now?

**Comment by pllim** [comment]
p.s. Not sure how to test this...

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/viewers.py:205]: When will the code ever enter here (not having already returned from the `self.compas is None`)?  I added a print statement and can't seem to find a scenario where its needed.

File: jdaviz/app.py
Comment (kecnry) [inline jdaviz/app.py:1176]: This does seem to work for the bug, but also has the consequence of re-enabling any intentionally disabled layers every time data is checked/unchecked/added.  This might be the best incremental solution for now, as I think the only way to avoid this would be to handle tracking the blinking state separately from the visible state.

**Review by kecnry — CHANGES_REQUESTED** [review]

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/viewers.py:190]: It is used in `app.py`. That's why I added it.

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/viewers.py:205]: The `@observe('plugin_opened')` on [_compass_with_new_viewer](https://github.com/spacetelescope/jdaviz/pull/1153/files#diff-d00ec66d428dae92765a78d062c61800d5c32561d2296f7f08b5d2058deca69eL41-R41) should handle re-drawing the compass as soon as the plugin is reopened (since it makes a call to `viewer.on_limits_change`).  It will then continue to redraw live from any limits changes if the plugin remains opened.  And the `if self.compass is None: return` at the top of `set_compass` will stop updating the compass while the plugin is closed (as all viewers have been set to have `viewer.compass = None` when the plugin was closed.

I don't think any of the `only_if_open` logic here is necessary at all, but if you find a case where that wasn't behaving as expected, I can look into it and try to fix it.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/viewers.py]: Yup, you are right. I removed the extra unnecessary code. Good catch!

**Comment by pllim** [comment]
@kecnry , since you said this works. Should we get this it first and revisit #1166 later?

**Comment by pllim** [comment]
@kecnry , since you said this works. Should we get this it first and revisit #1166 later?

File: jdaviz/app.py
Comment (kecnry) [reply jdaviz/app.py]: I'm still really torn about this.  Imviz supports both blinking and viewing multiple semi-transparent layers at once, but switching between those two "modes of operation" will still result in unexpected behavior.  That of course will take a lot more thought to fully "fix", and so I do think this is helpful for now.

Should we at least include a check that the layers we're forcing to be visible are data-layers and not regions, etc?  That's probably an edge case where a user would intentionally set a subset to be hidden, but at least this hard reset would affect as few layers as possible then.

**Review by rosteen — APPROVED** [review]
I think I'm ok with putting this fix in for now, since Kyle's PR is still in draft. I agree that this could have unintended consequences in terms of showing other things that should remain hidden, but my instinct is that those cases will come up less than the case this is fixing.

**Comment by pllim** [comment]
Would it make you feel better if I check this before resetting `.visible`?

https://github.com/spacetelescope/jdaviz/blob/be1d987905ef4f843982f0325457c2f091049a24/jdaviz/configs/imviz/helper.py#L469

**Comment by camipacifici** [comment]
Done a couple of tests loading two images and loading 4 images. In both cases, after blinking and deselecting the visible layer, the other (or another) layer comes up as expected.
I noticed that in the Plot option plugin, no layer is automatically selected in the "layer options" part, so half the plugin is not showing. Is this expected? I found it a little annoying. Not a show stopper, but to be considered.

**Comment by pllim** [comment]
Let's revisit this after #1178 is merged as reviewers might be confusing that bug with the work in this PR too.

**Comment by pllim** [comment]
> no layer is automatically selected in the "layer options" part

I think I know what you are saying but it only happens for me in a very specific workflow when I manually set the Layer under Layer Options. If I do not touch it, it seems to switch around smoothly. My workflow to reproduce what I think you are saying:

1. Load two images in Imviz.
2. Blink once to see the first one.
3. Blink again to see the second one.
4. Manually set the Layer in Layer Options to the second image (it automatically shows the first one).
5. Uncheck the second image being shown from Data dropdown.
6. Now Layer is not populated.

This also happens on `main` (together with a manifestation of *this* bug), so you could argue maybe fixing it is in scope but given that @kecnry is rethinking the whole Layer Options stuff, should we hold off on fixing it here and defer as future work?

**Review by kecnry — APPROVED** [review]
Definitely an improvement and probably worth the few odd behaviors it can introduce in more advanced usage until we get a chance to consider rewriting how blinking works entirely.

**Comment by pllim** [comment]
Let's tackle #1182 as new work because it needs more investigation and is definitely more than "1 point". 😬  Thanks for your understanding!

**Pull request merged**
